script/assemble_ember_build.rb

```diff
@@ -69,6 +79,44 @@ def existing_core_build_usable?
   end
 end
 
+def download_prebuild_assets!
+  return false if !DOWNLOAD_PRE_BUILT_ASSETS
+
+  git_is_clean = capture("git", "status", "--porcelain").strip.empty?
+  if !git_is_clean
+    log "Git working directory is not clean. Cannot download prebuilt assets."
+    return false
+  end
+
+  core_commit_hash = capture("git", "rev-parse", "HEAD").strip
+  version_string = "#{Discourse::VERSION::STRING}-#{core_commit_hash.slice(0, 8)}"
+
+  url = "#{PRE_BUILD_ROOT}/#{version_string}/#{resolved_ember_env}.tar.gz"
+  puts "Fetching and extracting #{url}..."
+
+  temp_file = "#{__dir__}/../tmp/assets.tar.gz"
+  begin
+    system("curl", "--fail-with-body", "-L", url, "-o", temp_file, exception: true)
+  rescue RuntimeError => e
+    log "Failed to download prebuilt assets: #{e.message}"
+    return false
+  end
+
+  FileUtils.rm_rf("dist")
+  FileUtils.mkdir_p("dist")
+  begin
+    system("tar", "--strip-components=1", "-xzf", temp_file, "-C", "dist", exception: true)
+  rescue RuntimeError => e
+    log "Failed to extract prebuilt assets: #{e.message}"
+    return false
+  end
+
+  FileUtils.rm_f(temp_file)
```

Comment:
I believe this needs to be called within an `ensure` block. Otherwise, it may not be cleaned up if we encounter a runtime error in any of the steps above.